src/decoder/plugins/VgmstreamDecoderPlugin.cxx

```diff
@@ -0,0 +1,332 @@
+	FmtDebug(vgmstream_domain, "vgmstream {}.{}.{}", major, minor, patch);
+
+	cfg.ignore_loop = block.GetBlockValue("ignore_loop", false);
+	cfg.force_loop = block.GetBlockValue("force_loop", false);
+	cfg.really_force_loop = block.GetBlockValue("really_force_loop", false);
+	cfg.ignore_fade = block.GetBlockValue("ignore_fade", false);
+
+	// same defaults as plugins shipped with vgmstream
+	VgmstreamGetDouble(block, &cfg.loop_count, "loop_count", 2.0);
+	VgmstreamGetDouble(block, &cfg.fade_time, "fade_time", 10.0);
+	VgmstreamGetDouble(block, &cfg.fade_delay, "fade_delay", 0.0);
+
+	return true;
+}
+
+static AudioFormat
+VgmstreamGetFormat(libvgmstream_t *lib, libstreamfile_t *file)
+{
+	uint32_t sample_rate = lib->format->sample_rate;
+	auto sample_format = SampleFormat::UNDEFINED;
+	uint8_t channels = lib->format->channels;
+
+	if (lib->format->sample_format == LIBVGMSTREAM_SFMT_PCM16) {
+		sample_format = SampleFormat::S16;
+	} else if (lib->format->sample_format == LIBVGMSTREAM_SFMT_PCM24) {
+		// mpd doesn't support S24 samples, so reopen the stream as S32
+		const int subsong = lib->format->subsong_index;
+
+		libvgmstream_close_stream(lib);
+
+		libvgmstream_config_t new_cfg = cfg;
+		new_cfg.force_sfmt = LIBVGMSTREAM_SFMT_PCM32;
+		libvgmstream_setup(lib, &new_cfg);
+
+		const int err = libvgmstream_open_stream(lib, file, subsong);
+		assert(err == 0);
+		assert(lib->format->sample_format == LIBVGMSTREAM_SFMT_PCM32);
+
+		sample_format = SampleFormat::S32;
+	} else if (lib->format->sample_format == LIBVGMSTREAM_SFMT_PCM32) {
+		sample_format = SampleFormat::S32;
+	} else if (lib->format->sample_format == LIBVGMSTREAM_SFMT_FLOAT) {
+		sample_format = SampleFormat::FLOAT;
+	}
+
+	return {sample_rate, sample_format, channels};
+}
+
+static int
+ParseSubsongName(const char *base) noexcept
+{
+	base = StringAfterPrefix(base, SUBSONG_PREFIX);
+	if (base == nullptr)
+		return 0;
+
+	char *endptr;
+	const auto track = strtol(base, &endptr, 10);
+	if (endptr == base || *endptr != '.')
+		return 0;
+
+	return static_cast<int>(track);
+}
+
+/**
+ * returns the file path stripped of any /song_xxx.* subsong suffix
+ * and the track number (or 0 if no "song_xxx" suffix is present).
+ */
+static VgmstreamContainerPath
+ParseContainerPath(Path path_fs)
+{
+	const Path base = path_fs.GetBase();
+	int track;
+	if (base.IsNull() || (track = ParseSubsongName(NarrowPath(base))) < 1)
+		return {AllocatedPath(path_fs), 0};
+
+	return {path_fs.GetDirectoryName(), track};
+}
+
+static bool
+VgmstreamScanTags(Path path, libvgmstream_t *lib, TagHandler &handler,
+		  ReplayGainInfo *rgi) noexcept
+{
+	bool found_replay_gain = false;
+	bool found_title = false;
+
+	const auto tags_path = AllocatedPath::Build(path.GetDirectoryName(), "!tags.m3u");
+	libstreamfile_t *tags_file = libstreamfile_open_from_stdio(tags_path.c_str());
+	if (tags_file == nullptr) {
+		if (*lib->format->stream_name)
+			handler.OnTag(TAG_TITLE, lib->format->stream_name);
```

Comment:
What's the point of returning a tag if this function fails?

Reply:
Most formats with subsongs supported by vgmstream will provide a human-readable identifier for each subsong. This can't be used as the filename, because it's not necessarily a *unique* identifier, but it's much nicer to show "BGM_SHOP" than "song_002.bank".

Reply:
That doesn't answer my question. This is the failure code path.

Reply:
vgmstream normally stores tags out-of-band in a separate file, but some supported formats have an internal name field. If you think it would make the code clearer I can try to find a way to separate that handling out.

Reply:
I think I don't understand at all what this piece of code is supposed to do. If the function fails (returns false), it must not invoke any handler methods. It's a bug.
So whatever you're trying to do, you're doing it wrong. But until I understand the purpose of this, I can't advise what else to do.

Reply:
Does the refactor I did make the intended logic clearer?